src/Titanium.Web.Proxy/Certificates/CertificateManager.cs

```diff
@@ -503,33 +507,50 @@ private X509Certificate2 makeCertificate(string certificateName, bool isRootCert
-            // This saves us from another expensive lock. The goal here is to minimize creation of multiple tasks for same certification name.
-            // Goal is not to guarantee single certificate creation, which is not needed in our case.
-            if (pendingCertificateCreationTasks.TryGetValue(certificateName, out var task))
-            {
-                return await task;
-            }
-
-            // run certificate creation task & add it to pending tasks         
-            task = Task.Run(() =>
-            {
-                var result = CreateCertificate(certificateName, false);
-                if (result != null)
+            var createdTask = false;
+            Task<X509Certificate2?> createCertificateTask;
+            await pendingCertificateCreationTaskLock.WaitAsync();
+            try {
+                // handle burst requests with same certificate name
+                // by checking for existing task for same certificate name
+                if (!pendingCertificateCreationTasks.TryGetValue(certificateName, out createCertificateTask))
                 {
-                    cachedCertificates.TryAdd(certificateName, new CachedCertificate(result));
-                }
-
-                return result;
-            });
-            pendingCertificateCreationTasks.TryAdd(certificateName, task);
+                    createdTask = true;
+                    // run certificate creation task & add it to pending tasks
+                    createCertificateTask = Task.Run(() =>
+                    {
+                        var result = CreateCertificate(certificateName, false);
+                        if (result != null)
+                        {
+                            cachedCertificates.TryAdd(certificateName, new CachedCertificate(result));
+                        }
 
-            // cleanup pending tasks & return result
-            var certificate = await task;
-            pendingCertificateCreationTasks.TryRemove(certificateName, out task);
+                        return result;
+                    });
+                    pendingCertificateCreationTasks[certificateName] = createCertificateTask;
+                }
+            } finally {
+                pendingCertificateCreationTaskLock.Release();
+            }
+
+            // create certificate outside of lock to allow other threads a chance to see the pending task.
+            // precludes the possiblility of multiple threads simultaneously generating the same certificate,
+            // we tolerate the case of a duplicate sequential generation due to a race (should be rare)
+            // t1: await createCertificateTask
+            // t2: cachedCertificates.TryGetValue->false
+            // t1: pendingCertificateCreationTasks.Remove
+            // t2: pendingCertificateCreationTasks.TryGetValue->false
+            var certificate = await createCertificateTask;
+
+            if (createdTask) {
```

Comment:
Do we need an extra lock acquire here? I think `pendingCertificateCreationTasks.TryRemove(certificateName)` will do the trick without `await pendingCertificateCreationTaskLock.WaitAsync()` and a `try finally` statement.